Multi Currency Improvements

PSzyszkaSpark opened the pull request:
Migrates [Spree Multi Currency](https://github.com/spree-contrib/spree_multi_currency) to Spree Core

- [X] adds the ability to set more than one currency per one Store (`supported_currencies` attribute)
- [X] adds the ability to set prices for multiple currencies for each Variant in the Admin Panel
- [X] adds the ability to set selected currency in API v2 via the `currency` URL param

PSzyszkaSpark committed:
Adds spree_multi_currency logic to spree 4.0

PSzyszkaSpark committed:
Adds spree_multi_currency logic to spree 4.0

PSzyszkaSpark committed:
Adds spree_multi_currency logic to spree 4.0

PSzyszkaSpark committed:
Corrects multi currency login

damianlegawiec commented on core/lib/spree/core/engine.rb:
```diff
@@ -142,6 +142,19 @@ class Engine < ::Rails::Engine
+
+      require 'spree/core/currency_helpers'
+
+      def self.activate
```
@PSzyszkaSpark we don't need the decorator activator code

damianlegawiec commented on core/lib/spree/core/engine.rb:
```diff
@@ -142,6 +142,19 @@ class Engine < ::Rails::Engine
+            Rails.configuration.cache_classes ? require(c) : load(c)
+          end
+        end
+        ApplicationController.send :include, Spree::CurrencyHelpers
```
@PSzyszkaSpark we should include this in Spree::BaseController

damianlegawiec commented on frontend/app/assets/javascripts/spree/frontend/multi_currency.js.coffee:
```diff
@@ -0,0 +1,9 @@
+$ ->
```
@PSzyszkaSpark let's replace the coffee file with plain JS please

damianlegawiec replied:
@PSzyszkaSpark we need this file in plain JS, not coffee, we've dropped coffee support in 3.7

damianlegawiec commented on frontend/spree_frontend.gemspec:
```diff
@@ -28,4 +28,5 @@ Gem::Specification.new do |s|
   s.add_dependency 'jquery-rails',    '~> 4.3'
 
   s.add_development_dependency 'capybara-accessible'
+  s.add_development_dependency 'coffee-rails'
```
@PSzyszkaSpark we don't use coffee script anymore

PSzyszkaSpark committed:
Fix for small issues

PSzyszkaSpark committed:
Corrects frontend currency change

PSzyszkaSpark committed:
Small spec fixes

damianlegawiec committed:
eslint fix

damianlegawiec commented on api/app/controllers/concerns/spree/api/v2/storefront/order_concern.rb:
```diff
@@ -30,10 +30,24 @@ def find_spree_current_order
           end
 
+          def use_current_currency
+            if session.key?(:currency) && supported_currencies.map(&:iso_code).include?(session[:currency])
```
@PSzyszkaSpark we don't use sessions in API endpoints

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    visit spree.edit_admin_general_settings_path
+
+    # Test initial state
+    expect(find_field('supported_currencies').value).to eq 'USD'
```
expect(page).to have_field('supported_currencies', with: 'USD')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+    # Test initial state
+    expect(find_field('supported_currencies').value).to eq 'USD'
+    expect(find('#allow_currency_change')).to_not be_checked
```
expect(page).to have_checked_field('allow_currency_change')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    # Test initial state
+    expect(find_field('supported_currencies').value).to eq 'USD'
+    expect(find('#allow_currency_change')).to_not be_checked
+    expect(find('#show_currency_selector')).to_not be_checked
```
expect(page).to have_unchecked_field('show_currency_selector')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+    # Interact with the form
+    fill_in 'supported_currencies', with: 'USD,PLN'
+    find(:css, '#allow_currency_change').set(true)
```
check('allow_currency_change')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    # Interact with the form
+    fill_in 'supported_currencies', with: 'USD,PLN'
+    find(:css, '#allow_currency_change').set(true)
+    find(:css, '#show_currency_selector').set(true)
```
check('show_currency_selector')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+    # Test final state
+    expect(page).to have_content 'General Settings has been successfully updated!'
+    expect(find_field('supported_currencies').value).to eq 'USD,PLN'
```
expect(page).to have_field('supported_currencies, with: 'USD,PLN')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    # Test final state
+    expect(page).to have_content 'General Settings has been successfully updated!'
+    expect(find_field('supported_currencies').value).to eq 'USD,PLN'
+    expect(find('#allow_currency_change')).to be_checked
```
expect(page).to have_checked_field('allow_currency_change')

twalpole commented on backend/spec/features/admin/general_settings/prices_settings_updating_spec.rb:
```diff
@@ -0,0 +1,34 @@
+    expect(page).to have_content 'General Settings has been successfully updated!'
+    expect(find_field('supported_currencies').value).to eq 'USD,PLN'
+    expect(find('#allow_currency_change')).to be_checked
+    expect(find('#show_currency_selector')).to be_checked
```
expect(page).to have_checked_field('show_currency_selector')

twalpole commented on frontend/spec/features/product_prices_spec.rb:
```diff
@@ -0,0 +1,58 @@
+
+      it 'will not render the currency selector' do
+        visit spree.product_path(product)
+        expect(page).to_not have_text 'Currency'
```
There should be a positive expectation prior to the negative expectation to ensure you're on the expected page

twalpole commented on frontend/spec/features/product_prices_spec.rb:
```diff
@@ -0,0 +1,58 @@
+
+        it 'will not render the currency selector' do
+          visit spree.product_path(product)
+          expect(page).to_not have_text 'Currency'
```
See above

twalpole reviewed (commented):
Using RSpec provided matchers (eq, be_xxx, etc) with Capybara returned elements opens you up to race conditions in the tests.  Prefer using the Capybara provided matchers instead;.

twalpole commented on frontend/spec/features/product_prices_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+      it 'will not render the currency selector' do
+        visit spree.product_path(product)
+        expect(current_path).to eq(spree.product_path(product))
```
Using the `eq` matcher here produces a race condition that can lead to flaky tests. Prefer Capybara provided matchers when they exist

    expect(page).to have_current_path(spree.product_path(product))

twalpole commented on frontend/spec/features/product_prices_spec.rb:
```diff
@@ -34,7 +34,7 @@
       it 'will not render the currency selector' do
         visit spree.product_path(product)
-        expect(current_path).to eq(spree.product_path(product))
+        expect(current_path).to have_current_path(spree.product_path(product))
```
The expectation needs to be on `page`  -   `expect(page).to have_current_path(...)`

twalpole commented on frontend/spec/features/order_prices_spec.rb:
```diff
@@ -19,7 +19,7 @@
       expect(page).to have_text '$19.99'
       select 'EUR', from: 'currency'
-      wait_for { page.have_text?('€16.00') }
+      wait_for { page.has_text?('€16.00') }
```
I've gone and looked further at this - and I believe the issue you're having is that the page before you press 'Add To Cart' already has the text '$19.99' on it - so you have a race condition where the currency can be selected during the Add To Cart action happening.  Instead you need to replace `expect(page).to have_text '$19.99'` with a check for something that isn't on the page until the 'Add To Cart' action has completed.  Something like

    expect(page).to have_css('#link-to-cart', text: '$19.99')

should do it (and then get rid of `wait_for` - it really should never be necessary to use)

PSzyszkaSpark replied:
Thanks for the advise @twalpole . To me it seems like a random error and even after waiting for 60s (which is very long), sometimes those specs didn't pass. That's why I decided to use `wait_for`. But I will try out your solution.

twalpole replied:
@PSzyszkaSpark The reason it was failing previously is because the selection of a new currency was occurring either while the `Add To Cart` action was being processed or before the page was ready to handle it, so no request to change the currency was actually done.  By changing to checking for the text within the specific element it ensures the action has been done and completed (the `link-to-cart` element is loaded via an XHR request after the add to cart action has been completed) so you know the currency selector is ready to be used.

PSzyszkaSpark replied:
Thank you for the explanation. Corrected as you suggested and it works perfectly:)

PSzyszkaSpark committed:
Removes change in spree_frontend.gemspec

PSzyszkaSpark committed:
Adds small fixes

PSzyszkaSpark committed:
Corrects line_item_spec.rb

PSzyszkaSpark committed:
Removes unnecessary test from product_spec.rb

damianlegawiec commented on frontend/Gemfile:
```diff
@@ -2,5 +2,6 @@ eval_gemfile('../common_spree_dependencies.rb')
 
 gem 'spree_core', path: '../core'
 gem 'spree_api', path: '../api'
+gem 'pry'
```
@PSzyszkaSpark we should remove this

damianlegawiec commented on frontend/app/assets/javascripts/spree/frontend/multi_currency.js:
```diff
@@ -0,0 +1,11 @@
+      data: {
+        currency: $(this).val()
+      }
+    }).done(() => window.location.reload())
```
@PSzyszkaSpark we cannot use arrow functions with the asset pipeline, we need to stick to good old functions

PSzyszkaSpark committed:
Corrects multi_currency.js and removes gem pry from frontend/Gemfile

PSzyszkaSpark committed:
Deletes multi_currency.js.coffee file

damianlegawiec committed:
Streamline and unify currency control

damianlegawiec committed:
Test API v2 Cart endpoint against different currencies

btolarz committed:
Moved supported currencies from preferencies to store

damianlegawiec committed:
fix

damianlegawiec committed:
Make supported_currrencies optional for Store model

If not set fallback to default_currency value. This will streamline the migration process for existing applications.

damianlegawiec committed:
Fixed setting supported currencies in Store Admin Panel

damianlegawiec committed:
Fixed API v1 stores endpoint spec

damianlegawiec committed:
Fix setting currency via URL param for API v2

damianlegawiec committed:
Fixes Storefront specs

damianlegawiec committed:
Upgrade Admin Prices UI
